Sprint 8 solution time and duration

Altron2108 opened the pull request.

Altron2108 committed:
Добавлен метод `updateEpicTime` для пересчета времени начала, окончания и длительности эпика на основе подзадач.
- Добавлены методы для работы с подзадачами: `addSubtask`, `removeSubtask`, `clearSubtasks`.
- Реализован метод `toString` для удобного вывода информации об эпике, включая время начала, окончания и длительность.
- Обновлены поля `duration` и `endTime` для учета изменений времени и длительности эпика.

Altron2108 committed:
- Реализованы тесты для проверки статуса эпика в различных ситуациях:
  - `testEpicStatusWhenNoSubtasks`: проверка статуса NEW, когда подзадачи отсутствуют.
  - `testEpicStatusWithAllNewSubtasks`: проверка статуса NEW, когда все подзадачи NEW.
  - `testEpicStatusWithAllDoneSubtasks`: проверка статуса DONE, когда все подзадачи DONE.
  - `testEpicStatusWithNewAndInProgressSubtasks`: проверка статуса IN_PROGRESS, когда подзадачи имеют смешанные статусы.
  - `testEpicStatusWithAllInProgressSubtasks`: проверка статуса IN_PROGRESS, когда все подзадачи IN_PROGRESS.

Altron2108 committed:
- Добавлен класс FileBackedTaskManager, наследующий от InMemoryTaskManager.
- Реализованы методы для сохранения задач в файл (save) и загрузки задач из файла (load).
- Сериализация включает в себя новые поля: id, type, name, status, description, duration, startTime, epicId.
- Добавлены методы для преобразования задач в строку (toString) и обратно в объект Task (fromString).
- Обработка разных типов задач (RegularTask, Epic, Subtask) при загрузке из файла.

Altron2108 committed:
- Реализован тест testSaveAndLoadTasks для проверки сохранения задачи в файл и её последующей загрузки.
- Проверяется корректность загруженной задачи: наличие задачи в менеджере и соответствие загруженных значений оригинальным.
- Используется временный файл "test_tasks.csv" для хранения тестовых данных.

Altron2108 committed:
- Реализован тест addTaskToHistory_TaskAdded_HistoryContainsTask для проверки добавления задачи в историю.
- Проверяется, что после добавления задачи история не пуста, содержит одну задачу и эта задача соответствует добавленной.
- Используется класс RegularTask для создания тестовой задачи.

Altron2108 committed:
- Реализован интерфейс HistoryManager с поддержкой хранения и управления историей просмотров задач.
- Используется двусвязный список для обеспечения эффективного добавления и удаления задач.
- Реализованы методы добавления задачи (add), удаления задачи (remove) и получения истории задач (getHistory).
- Задачи хранятся в виде узлов (Node) с указателями на предыдущий и следующий элементы.

Altron2108 committed:
- Реализованы тесты для проверки функциональности менеджера истории задач:
  - `addTasksToHistory_HistoryPreservesCorrectOrder`: проверяет сохранение порядка задач в истории.
- `addSameTaskToHistory_TaskMovesToEnd`: проверяет, что повторно добавленная задача перемещается в конец истории.
- `addDuplicateTask_HistoryDoesNotContainDuplicates`: проверяет, что история не содержит дубликатов.
- `getHistory_HistoryReturnsUnmodifiableList`: проверяет, что история возвращает не модифицируемый список.
- `removeTaskFromHistory_TaskIsRemoved`: проверяет удаление задачи из истории.
- `removeNonExistentTask_HistoryRemainsUnchanged`: проверяет, что удаление несуществующей задачи не меняет историю.
- `getHistory_WhenNoTasksAdded_ReturnsEmptyList`: проверяет, что история пуста, если не было добавлено задач.

Altron2108 committed:
- Реализован класс InMemoryTaskManager, который управляет задачами, эпиками и подзадачами в памяти.
- Добавлены методы для создания, получения, обновления и удаления задач, эпиков и подзадач.
- Реализована валидация пересечений задач по времени перед созданием и обновлением.
- Добавлены методы для обновления времени и статуса эпиков на основе подзадач.
- Используется TreeSet для хранения задач в приоритетном порядке на основе времени начала.
- Реализованы методы получения всех задач, эпиков и подзадач.

Altron2108 committed:
- Реализованы тесты для проверки создания задач с учетом пересечений по времени.
- Тесты включают проверку:
- Успешного создания задач без пересечений.
- Генерации исключений при создании задач с пересечениями.
- Генерации исключений при обновлении задач с пересечениями.
- Правильного порядка задач при получении приоритетных задач.
- Использованы JUnit для организации тестов и проверки условий.

Altron2108 committed:
!!!

Altron2108 committed:
- Создан класс Main для демонстрации функциональности менеджера задач.
- Реализовано создание обычных задач (RegularTask) и эпиков (Epic).
- Добавлены подзадачи (Subtask) с указанием времени начала и длительности.
- Внедрена история просмотров задач через HistoryManager.
- Реализовано удаление задач и эпиков с обновлением истории.
- Включен вывод состояния истории после различных операций для проверки функциональности.

Altron2108 committed:
- Реализован класс RegularTask, наследующий от Task.
- Добавлен конструктор для инициализации задачи с параметрами: заголовок, описание, статус, длительность и время начала.
- Переопределён метод getType() для возвращения типа задачи.
- Переопределён метод toString() для удобного представления информации о задаче, включая идентификатор, заголовок, описание, статус, длительность и время начала.

Altron2108 committed:
- Реализован класс Subtask, наследующий от Task.
- Добавлен конструктор для инициализации подзадачи с параметрами: заголовок, описание, статус, длительность, время начала и идентификатор эпика.
- Добавлен метод getEpicId() для получения идентификатора эпика, к которому принадлежит подзадача.
- Переопределён метод getType() для возвращения типа подзадачи.
- Переопределён метод toString() для удобного представления информации о подзадаче, включая идентификатор, заголовок, описание, статус, длительность, время начала и идентификатор эпика.

Altron2108 committed:
- Реализован тест constructor_ShouldInitializeFieldsCorrectly для проверки корректной инициализации полей подзадачи.
- Проверяются значения заголовка, описания, статуса и идентификатора эпика.
- Используются ассерты для проверки правильности инициализации полей при создании объекта Subtask.

Altron2108 committed:
- Убраны модификаторы final для полей title, description и status, что позволяет их изменять.
- Добавлены сеттеры для полей title, description и status.
- Установлен идентификатор id по умолчанию равным -1.
- Реализован метод getEndTime() для вычисления времени завершения задачи на основе ее длительности.
- Структура класса улучшена для повышения гибкости и удобства использования.

Altron2108 committed:
- Реализованы тесты для проверки начальных значений полей title, description и status.
- Использован класс RegularTask для тестирования его свойств.
- Добавлены сообщения к тестам для лучшего понимания их назначения.

Altron2108 committed:
- Добавлены методы для создания, получения, обновления и удаления задач, эпиков и подзадач.
- Включены методы для удаления всех задач и получения списков всех задач, эпиков и подзадач.
- Добавлен метод для получения задач по приоритету.

Altron2108 committed:
- Реализован тест createTask_ShouldAssignUniqueIds для проверки, что задачи получают уникальные идентификаторы.
- Используется фиксированное время для обеспечения согласованности тестов.
- Проверяется, что два создаваемых задания имеют разные идентификаторы.

Altron2108 committed:
- Реализован тест taskConstructor_ShouldInitializeFieldsCorrectly для проверки корректной инициализации полей при создании задачи.
- Добавлен тест setTitle_setDescription_setStatus_ShouldUpdateTaskFields для проверки обновления заголовка, описания и статуса задачи.
- Используются ассерты для проверки соответствия значений после изменений.

Altron2108 committed:
!!

Altron2108 committed:
- Реализован тест testSaveAndLoadTasks, проверяющий корректное сохранение и загрузку задач из файла.
- Добавлено наследование от абстрактного класса TaskManagerTest для унификации тестов.
- Обновлен метод setUp для инициализации FileBackedTaskManager перед каждым тестом.

Altron2108 committed:
- Реализован набор тестов для проверки корректности создания задач без пересечений.
- Добавлен тест для проверки исключений при попытке создания задач с пересечением времени.
- Реализован тест для проверки исключений при обновлении задач с пересечением времени.
- Добавлен тест для проверки правильного порядка возвращаемых задач по приоритету.
- Обновлено наследование от абстрактного класса TaskManagerTest для унификации тестов.

Altron2108 committed:
- Создан абстрактный класс TaskManagerTest для общего функционала тестирования.
- Реализованы тесты для проверки создания, обновления и удаления задач.
- Добавлены проверки на уникальность идентификаторов задач.
- Реализованы тесты для проверки удаления задач по идентификатору и удаления всех задач.
- Включены тесты для получения всех задач и проверки их наличия.

Altron2108 committed:
- Создан абстрактный класс TaskManagerTest для общего функционала тестирования.
- Реализованы тесты для проверки создания, обновления и удаления задач.
- Добавлены проверки на уникальность идентификаторов задач.
- Реализованы тесты для проверки удаления задач по идентификатору и удаления всех задач.
- Включены тесты для получения всех задач и проверки их наличия.

Altron2108 committed:
!!

Altron2108 committed:
- Тестирование создания задач с уникальными идентификаторами.
- Тестирование удаления задачи по идентификатору.
- Тестирование получения всех задач.
- Тестирование обновления задачи.
- Тестирование удаления всех задач.
- Тестирование метода getPrioritizedTasks для проверки сортировки задач по приоритету на основе времени начала.

Altron2108 committed:
!!

Altron2108 committed:
- Тесты проверки отсутствия пересечения времени задач (testTaskWithoutOverlap, testTaskWithOverlapThrowsException)
- Тест на обновление задачи с пересечением времени (testUpdateTaskWithOverlapThrowsException)
- Тест на получение приоритезированных задач (getPrioritizedTasks_ShouldReturnTasksInOrder)

Altron2108 committed:
Добавлены тесты для проверки пересечения времени задач в InMemoryTaskManager:
- testTaskWithoutOverlap: проверка создания задач без пересечения по времени.
- testTaskWithOverlapThrowsException: проверка выброса исключения при создании задач с пересечением по времени.
- testUpdateTaskWithOverlapThrowsException: проверка выброса исключения при обновлении задачи с пересечением по времени.

Altron2108 committed:
Добавлен тест для проверки сохранения и загрузки задач в FileBackedTaskManager:
- testSaveAndLoadTasks: проверка корректного сохранения задачи в файл и загрузки из файла.
- getPrioritizedTasks_ShouldReturnTasksInOrder: проверка возвращения задач в порядке приоритета по времени начала.

Altron2108 committed:
Добавлен тест для проверки сохранения и загрузки задач в FileBackedTaskManager:
- testSaveAndLoadTasks: проверка корректного сохранения задачи в файл и загрузки из файла.
- getPrioritizedTasks_ShouldReturnTasksInOrder: проверка возвращения задач в порядке приоритета по времени начала.

Altron2108 committed:
Добавлен тест для проверки сохранения и загрузки задач в FileBackedTaskManager:
- testSaveAndLoadTasks: проверка корректного сохранения задачи в файл и загрузки из файла.
- getPrioritizedTasks_ShouldReturnTasksInOrder: проверка возвращения задач в порядке приоритета по времени начала.

Altron2108 committed:
Добавлен тест для проверки сохранения и загрузки задач в FileBackedTaskManager:
- testSaveAndLoadTasks: проверка корректного сохранения задачи в файл и загрузки из файла.
- getPrioritizedTasks_ShouldReturnTasksInOrder: проверка возвращения задач в порядке приоритета по времени начала.

Altron2108 committed:
!!!

Altron2108 committed:
!!!

Altron2108 committed:
!!!

Altron2108 committed:
!!!

Altron2108 committed:
!!!

Altron2108 committed:
!!!

Altron2108 committed:
!!!

johnskyi commented on src/FileBackedTaskManager.java:
```diff
@@ -1,195 +1,82 @@
-    // Метод для сохранения состояния менеджера в файл
-    private void save() {
+    // Сериализация задач
+    public void save() throws IOException {
```
зачем этот метод сделал публичным?  он нигде кроме этог окласса не должен использоваться

johnskyi commented on src/FileBackedTaskManager.java:
```diff
@@ -1,195 +1,82 @@
-        return id;
-    }
-
-    @Override
```
а этот функционал почему удален?

johnskyi commented on test/EpicTest.java:
```diff
@@ -1,21 +1,80 @@
+    }
+
+    @Test
+    public void testEpicStatusWhenNoSubtasks() {
```
Есть соглашение по наименованию тестовых методов. В них нельзя использовать слово test, а принято называть по схеме какойМетодТестируемОжидаемыйРезультат

Altron2108 committed:
- Добавлен вызов метода `save()` в методах `updateTask()`, `updateEpic()`, `updateSubtask()`, `deleteTaskById()`, `deleteEpicById()` и `deleteSubtaskById()` для автоматического сохранения данных после изменений.
- Обработка исключений `IOException` с помощью `fillInStackTrace()` для сохранения стека вызовов в случае ошибок при сохранении данных в файл.

Altron2108 committed:
- Добавлен тест для сохранения и загрузки задач из файла (saveAndLoadTasks_WhenTaskIsCreated_ShouldSaveAndLoadTasksCorrectly).
- Добавлен тест для проверки сортировки задач по приоритету (getPrioritizedTasks_ShouldReturnTasksInOrder).
- Обновлен метод setUp() для инициализации FileBackedTaskManager с тестовым файлом.

Altron2108 committed:
- Реализован метод `save()` для сохранения задач в файл в формате CSV.
- Добавлен метод `load()` для загрузки задач из файла.
- Реализовано преобразование задачи в строку через `toString()` и обратное преобразование через `fromString()`.
- Добавлены вызовы метода `save()` в методы обновления и удаления задач, эпиков и подзадач, чтобы автоматически сохранять изменения в файл.
- Исправлена обработка ошибок при сохранении задач.

Altron2108 committed:
- Тест `shouldSaveAndLoadTasksCorrectly`: проверка корректного сохранения и загрузки задач из файла.
- Тест `getPrioritizedTasks_ShouldReturnTasksInOrder`: проверка возвращения задач в порядке приоритета.

Altron2108 committed:
- Тест `shouldSaveAndLoadTasksCorrectly`: проверка корректного сохранения и загрузки задач из файла.
- Тест `getPrioritizedTasks_ShouldReturnTasksInOrder`: проверка возвращения задач в порядке приоритета.
Добавлен метод tearDown для удаления тестового файла после каждого теста.

Altron2108 committed:
!!!

Altron2108 committed:
- Реализованы тесты для проверки статуса эпика в зависимости от статусов подзадач:
- Проверка, что статус эпика NEW, если нет подзадач.
- Проверка, что статус эпика NEW, если все подзадачи имеют статус NEW.
- Проверка, что статус эпика DONE, если все подзадачи имеют статус DONE.
- Проверка, что статус эпика IN_PROGRESS, если подзадачи имеют смешанные статусы.
- Проверка, что статус эпика IN_PROGRESS, если все подзадачи имеют статус IN_PROGRESS.
